File: Python/Easy/2942_find_words_containing_character.py
"""
2942. Find Words Containing Character

Given a 0-indexed array of strings 'words' and a character 'x',
return an array of indices representing the words that contain
the character 'x'.

Note that the returned array may be in any order.

Problem Link: https://leetcode.com/problems/find-words-containing-character/

Example 1:
    Input: words = ["leet","code"], x = "e"
    Output: [0, 1]
    Explanation: "e" occurs in both words: "leet", and "code". Hence, we
    return indices 0 and 1.

Example 2:
    Input: words = ["abc","bcd","aaaa","cbc"], x = "a"
    Output: [0, 2]
    Explanation: "a" occurs in "abc", and "aaaa". Hence, we return indices
    0 and 2.

Example 3:
    Input: words = ["abc","bcd","aaaa","cbc"], x = "z"
    Output: []
    Explanation: "z" does not occur in any of the words. Hence, we return
    an empty array.

Constraints:
    - 1 <= words.length <= 50
    - 1 <= words[i].length <= 50
    - x is a lowercase English letter.
    - words[i] consists only of lowercase English letters.
"""
from typing import List


class Solution:

    def findWordsContaining(self, words: List[str], x: str) -> List[int]:
        """
        This method finds all the indices of words in the given list that
        contain the character 'x'.

        Time Complexity: O(n * m), where n is the number of words and m is
        the average length of the words.
        Space Complexity: O(k), where k is the number of words containing
        the character x.

        Args:
        words (List[str]): List of words.
        x (str): The character to check for in each word.

        Returns:
        List[int]: List of indices of words containing the character 'x'.
        """
        wordCount: List[int] = []
        for pos, w in enumerate(words):
            if x in w:
                wordCount.append(pos)
        return wordCount


if __name__ == "__main__":
    solution = Solution()

    # Test cases
    test_cases = [
        (["leet", "code"], "e", [0, 1]),        # Test case 1
        (["abc", "bcd", "aaaa", "cbc"], "a", [0, 2]),  # Test case 2
        (["abc", "bcd", "aaaa", "cbc"], "z", []),     # Test case 3
    ]

    for i, (words, x, expected) in enumerate(test_cases, 1):
        result = solution.findWordsContaining(words, x)
        assert sorted(result) == sorted(expected), f"""
        Test case {i} failed: expected {expected}, got {result}"""
        print(f"Test case {i} passed!")

    print("All test cases passed!")

File: Golang/Easy/2942_find_words_containing_character.go
package main

import (
	"fmt"
	"strings"
)

// findWordsContaining returns every index i such that words[i] contains the
// character x.
//
// strings.Contains handles the substring search. Order matches input order.
//
// Time complexity:  O(n * m), n = len(words), m = avg word length.
// Space complexity: O(k), k = number of matching words.
func findWordsContaining(words []string, x byte) []int {
	indexes := []int{}
	for i, word := range words {
		if strings.Contains(word, string(x)) {
			indexes = append(indexes, i)
		}
	}

	return indexes
}

// findWordsContainingBare is the manual byte-scan variant — no stdlib search,
// breaks on first match per word.
//
// Time complexity:  O(n * m)
// Space complexity: O(k)
func findWordsContainingBare(words []string, x byte) []int {
	res := []int{}

	for i, word := range words {
		for j := 0; j < len(word); j++ {
			if word[j] == x {
				res = append(res, i)
				break
			}
		}
	}
	return res
}

func main() {
	testCases := []struct {
		words    []string
		x        byte
		expected []int
	}{
		{[]string{"leet", "code"}, 'e', []int{0, 1}},
		{[]string{"abc", "bcd", "aaaa", "cbc"}, 'a', []int{0, 2}},
		{[]string{"abc", "bcd", "aaaa", "cbc"}, 'z', []int{}},
	}

	funcs := []struct {
		label string
		fn    func([]string, byte) []int
	}{
		{"findWordsContaining    ", findWordsContaining},
		{"findWordsContainingBare", findWordsContainingBare},
	}

	for _, f := range funcs {
		for i, tc := range testCases {
			result := f.fn(tc.words, tc.x)
			status := "PASS"
			if fmt.Sprintf("%v", result) != fmt.Sprintf("%v", tc.expected) {
				status = "FAIL"
			}
			fmt.Printf("%s | Example %d | Input: words=%v x=%q | Expected: %v | Result: %v | [%s]\n",
				f.label, i+1, tc.words, tc.x, tc.expected, result, status)
		}
	}
}
